apollo-router/src/plugins/authentication.rs

```diff
@@ -346,120 +494,98 @@ impl Plugin for AuthenticationPlugin {
                         }
                     };
 
-                    // Try to find the kid of the header
-                    let kid = match jwt_header.kid {
-                        Some(k) => k,
-                        None => {
+                    // Extract our search criteria from our jwt
+                    let criteria = JWTCriteria {
+                        kid: jwt_header.kid,
+                        alg: jwt_header.alg,
+                    };
+
+                    // Search our set of JWKS to find the kid and process it
+                    // Note: This will search through JWKS in the order in which they are defined
+                    // in configuration.
+
+                    let jwk_opt = match search_jwks(my_jwks.clone(), &criteria, my_jwks_urls, &request.context).await {
+                        Ok(j) => j,
+                        Err(e) => {
                             return failure_message(
                                 request.context,
-                                "Missing kid value from JWT header".to_string(),
-                                StatusCode::BAD_REQUEST,
+                                format!("Could not retrieve JWKS set: {e}"),
+                                StatusCode::INTERNAL_SERVER_ERROR, // XXX: Best error?
```

Comment:
I think a 401 would be better here, as it's still authentication related. A 500 would work better if we cannot download the JWKS

Reply:
That is what is happening in this case. If we have sets but can't find a suitable match, that is dealt with in COOLDOWN processing where we will return an UNAUTHORIZED.